<html>
<head>
    <title>rnaSPAdes manual</title>
    <style type="text/css">
        .code {
            background-color: lightgray;
        }
    </style>
</head>
<body>
<h1>rnaSPAdes manual</h1>

1. <a href="#sec1">About rnaSPAdes</a><br>
2. <a href="#sec2">rnaSPAdes specifics</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;2.1. <a href="#sec2.1">Running rnaSPAdes</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;2.2. <a href="#sec2.2">rnaSPAdes-specific options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;2.2. <a href="#sec2.3">Assemblying strand-specific RNA-Seq</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;2.3. <a href="#sec2.4">rnaSPAdes output</a><br>
3. <a href="#sec3">Assembly evaluation</a><br>
4. <a href="#sec4">Citation</a><br>
5. <a href="#sec5">Feedback and bug reports</a><br>

<a name="sec1"></a>
<h2>1 About rnaSPAdes</h2>

<p> rnaSPAdes is a tool for <i>de novo</i> transcriptome assembly from RNA-Seq data and is suitable for all kind of organisms. rnaSPAdes is a part of <a href="http://cab.spbu.ru/software/spades/" target="_blank">SPAdes package</a> since version 3.9. Information about SPAdes download, requirements, installation and basic options can be found in <a href="manual.html" target="_blank">SPAdes manual</a>. Below you may find information about differences between SPAdes and rnaSPAdes.

<a name="sec2"></a>
<h2>2 rnaSPAdes specifics</h2>

<a name="sec2.1"></a>
<h3>2.1 Running rnaSPAdes</h3>
<p>
To run rnaSPAdes use

<pre class="code">
<code>
    rnaspades.py [options] -o &lt;output_dir>
</code>
</pre>

or 

<pre class="code">
<code>
    spades.py --rna [options] -o &lt;output_dir>
</code>
</pre>

Note that we assume that SPAdes installation directory is added to the <code>PATH</code> variable (provide full path to rnaSPAdes executable otherwise: <code>&lt;rnaspades installation dir>/rnaspades.py</code>). 

<p>Here are several notes regarding rnaSPAdes options:
    <ul>
        <li>rnaSPAdes can take as an input only paired-end and single-end libraries.</li>
        <li>rnaSPAdes does not support <code>--careful</code> and <code>--cov-cutoff</code> options.</li>
        <li>rnaSPAdes is not compatible with other pipeline options such as <code>--meta</code>, <code>--sc</code> and <code>--plasmid</code>. If you wish to assemble metatranscriptomic data just run rnaSPAdes as it is.</li>
        <li>By default rnaSPAdes uses 2 k-mer sizes, which are automatically detected using read length (approximately one third and half of the maximal read length). We recommend not to change this parameter because smaller k-mer sizes typically result in multiple chimeric (misassembled) transcripts. In case you have any doubts about your run, do not hesitate to contact us using e-mail given below.</li>
    </ul>

<a name="sec2.2"></a>
<h3>2.2 rnaSPAdes-specific options</h3>
<p>
    <code>--fast</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Uses only a single k-mer size (detected automatically) and removes short low-covered isolated edges from the graph. Note, that very short and low-expressed transcripts may be missing when this option is used.
</p>

<a name="sec2.3"></a>
<h3>2.3 Assemblying strand-specific RNA-Seq</h3>
<p>rnaSPAdes supports strand-specific RNA-Seq dataset. You can indicate that the dataset is strand-specific using one of the following options:

<p>
    <code>--ss-fr</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;The data set is strand-specific and first read in pair corresponds to <b>actual</b> gene strand.
</p>

<p>
    <code>--ss-rf</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;The data set is strand-specific and first read in pair corresponds to <b>reverse</b> gene strand (antisense).
</p>
<p>
Note, that strand-specificity is not realated and should not be confused with FR and RF orientation of paired reads. RNA-Seq paired-end reads typically have forward-reverse orientation (--> &lt;--), which is assumed by default and no additional options are needed (see <a href="manual.html#inputdata" target="_blank">main manual</a> for deatails).
<p>
If the data set is single-end use <code>--ss-fr</code> option in case when reads correspond to gene strand and <code>--ss-rf</code> otherwise.



<a name="sec2.4"></a>
<h3>2.4 rnaSPAdes output</h3>
<p>
rnaSPAdes outputs one main FASTA file named <code>transcripts.fasta</code>. The corresponding file with paths in the <code>assembly_graph.fastg</code> is <code>transcripts.paths</code>.

<p> 
In addition rnaSPAdes outputs transcripts with different level of filtration into <code>&lt;output_dir>/</code>: <br>
    <ul>
        <li><code>hard_filtered_transcripts.fasta</code> &ndash; includes only long and reliable transcripts with rather high expression.</li>
        <li><code>soft_filtered_transcripts.fasta</code> &ndash; includes short and low-expressed transcipts, likely to contain junk sequences.</li>
    </ul>
We reccomend to use main <code>transcripts.fasta</code> file in case you don't have any specific needs for you projects. Do not hesitate to contact us using e-mail given below.

<p>
   Contigs/scaffolds names in rnaSPAdes output FASTA files have the following format: <br><code>>NODE_97_length_6237_cov_11.9819_g8_i2</code><br> Similarly to SPAdes, <code>97</code> is the number of the transcript, <code>6237</code> is its sequence length in nucleotides and <code>11.9819</code> is the k-mer coverage. Note that the k-mer coverage is always lower than the read (per-base) coverage. <code>g8_i2</code> correspond to the gene number 8 and isoform number 2 within this gene. Transcripts with the same gene number are presumably received from same or somewhat similar (e.g. paralogous) genes. Note, that the prediction is based on the presence of shared sequences in the transcripts and is very approximate.

<a name="sec3">
<h2>3 Assembly evaluation</h2>

<p>
    <a href="http://cab.spbu.ru/software/rnaquast/" target="_blank">rnaQUAST</a> may be used for transcriptome assembly quality assessment for model organisms when reference genome and gene database are available. rnaQUAST also includes  <a href="http://busco.ezlab.org/" target="_blank">BUSCO</a> and <a href="http://topaz.gatech.edu/GeneMark/" target="_blank"> GeneMarkS-T</a> tools for <i>de novo</i> evaluation.
<br>

<a name="sec4">
<h2>4 Citation</h2>
<p>
If you use rnaSPAdes in your research, please include main SPAdes paper <a href="http://online.liebertpub.com/doi/abs/10.1089/cmb.2012.0021" target="_blank">Bankevich, Nurk et al., 2012</a> in your reference list. Paper on rnaSPAdes is to be submitted.


<a name="sec5">
<h2>5 Feedback and bug reports</h2>
<p>
    For rnaSPAdes citation use <a href="https://www.biorxiv.org/content/early/2018/09/18/420208" target="_blank">Bushmanova et al., 2018</a>.

<p>
    Your comments, bug reports, and suggestions are very welcomed. They will help us to further improve rnaSPAdes.
    If you have any troubles running rnaSPAdes, please send us <code>params.txt</code> and <code>spades.log</code> from the directory <code>&lt;output_dir></code>.

<p>
    Address for communications: <a href="mailto:spades.support@cab.spbu.ru" target="_blank">spades.support@cab.spbu.ru</a>.

<br/><br/><br/><br/><br/>

</body>
</html>
